pkg/ipam/poolallocator/allocator.go

```diff
@@ -0,0 +1,254 @@
+func NewIPPoolAllocator(poolName string, client crdclientset.Interface) (*IPPoolAllocator, error) {
+
+	allocator := &IPPoolAllocator{
+		ipPoolName: poolName,
+		crdClient:  client,
+	}
+	return allocator, nil
+}
+
+// initAllocatorList reads IP Pool status and initializes a list of allocators based on
+// IP Pool spec and state of allocation recorded in the status
+func (a *IPPoolAllocator) initIPAllocators(ipPool *v1alpha2.IPPool) (ipallocator.MultiIPAllocator, error) {
+
+	var allocators ipallocator.MultiIPAllocator
+
+	// Initialize a list of IP allocators based on pool spec
+	for _, ipRange := range ipPool.Spec.IPRanges {
+		if len(ipRange.CIDR) > 0 {
+			allocator, err := ipallocator.NewCIDRAllocator(ipRange.CIDR)
+			if err != nil {
+				return allocators, err
+			}
+			allocators = append(allocators, allocator)
+		} else {
+			allocator, err := ipallocator.NewIPRangeAllocator(ipRange.Start, ipRange.End)
+			if err != nil {
+				return allocators, err
+			}
+			allocators = append(allocators, allocator)
+		}
+	}
+
+	// Mark allocated IPs from pool status as unavailable
+	for _, ip := range ipPool.Status.Usage {
+		err := allocators.AllocateIP(net.ParseIP(ip.IPAddress))
+		if err != nil {
+			// TODO - fix state if possible
+			return allocators, fmt.Errorf("Inconsistent state for IP Pool %s with IP %s", ipPool.Name, ip.IPAddress)
+		}
+	}
+
+	return allocators, nil
+}
+
+func (a *IPPoolAllocator) readPoolAndInitIPAllocators() (*v1alpha2.IPPool, ipallocator.MultiIPAllocator, error) {
+	ipPool, err := a.crdClient.CrdV1alpha2().IPPools().Get(context.TODO(), a.ipPoolName, metav1.GetOptions{})
+
+	if err != nil {
+		return nil, ipallocator.MultiIPAllocator{}, err
+	}
+
+	allocators, err := a.initIPAllocators(ipPool)
+	if err != nil {
+		return nil, ipallocator.MultiIPAllocator{}, err
+	}
+	return ipPool, allocators, nil
+}
+
+func (a *IPPoolAllocator) appendPoolUsage(ipPool *v1alpha2.IPPool, ip net.IP, state v1alpha2.IPPoolUsageState, resource string) error {
+	usageEntry := v1alpha2.IPPoolUsage{
+		IPAddress: ip.String(),
+		State:     state,
+		Resource:  resource,
+	}
+
+	newPool := ipPool.DeepCopy()
+	newPool.Status.Usage = append(ipPool.Status.Usage, usageEntry)
+	_, err := a.crdClient.CrdV1alpha2().IPPools().UpdateStatus(context.TODO(), newPool, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Warningf("IP Pool %s update failed: %+v", ipPool.Name, err)
+		return err
+	}
+	klog.Infof("IP Pool update successful %s: %+v", ipPool.Name, newPool.Status)
+	return nil
+
+}
+
+func (a *IPPoolAllocator) removePoolUsage(ipPool *v1alpha2.IPPool, ip net.IP) error {
+
+	ipString := ip.String()
+	var newList []v1alpha2.IPPoolUsage
+	for _, entry := range ipPool.Status.Usage {
+		if entry.IPAddress != ipString {
+			newList = append(newList, entry)
+		}
+	}
+
+	if len(newList) == len(ipPool.Status.Usage) {
+		return fmt.Errorf("IP address %s was not allocated from IP pool %s", ip, ipPool.Name)
+	}
+
+	ipPool.Status.Usage = newList
```

Comment:
as ipPool is a pointer, would a copy be needed, as done in appendPoolUsage?

Reply:
I actually think deep copy is not needed in both, as appendPoolUsage and removePoolUsage is the last thing we do with the pool. But perhaps its safer to leave the copy in both in case function changes in future?

Reply:
In the future these functions could be called by other places. Since the copy isn't very expensive, I'd do it for bhoth append and remove.